// angular-hashtable/fesm2015/angular-hashtable.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
/**
 * @template T, L
 */
class HashTable {
    constructor() {
        this.table = {};
    }
    /**
     * @param {?} key
     * @param {?} value
     * @return {?}
     */
    put(key, value) {
        this.table['v_' + key] = value;
        return this;
    }
    /**
     * @param {?} key
     * @return {?}
     */
    get(key) {
        return this.table['v_' + key];
    }
    /**
     * @return {?}
     */
    getAll() {
        const /** @type {?} */ vector = Array.from(Object.keys(this.table), k => this.table[k]);
        return vector;
    }
    /**
     * @return {?}
     */
    getKeys() {
        const /** @type {?} */ keys = Array.from(Object.keys(this.table), k => k.substring(2));
        return keys;
    }
    /**
     * @param {?} key
     * @return {?}
     */
    has(key) {
        if (typeof this.table['v_' + key] !== 'undefined') {
            return true;
        }
        return false;
    }
    /**
     * @param {?} key
     * @return {?}
     */
    remove(key) {
        delete this.table['v_' + key];
        return this;
    }
    /**
     * @param {?} key
     * @param {?} value
     * @return {?}
     */
    putArray(key, value) {
        if (typeof this.table['a_' + key] === 'undefined') {
            this.table['a_' + key] = [];
        }
        this.table['a_' + key].push(value);
        return this;
    }
    /**
     * @param {?} key
     * @return {?}
     */
    getArray(key) {
        if (typeof this.table['a_' + key] === 'undefined') {
            this.table['a_' + key] = [];
        }
        return this.table['a_' + key];
    }
    /**
     * @param {?} key
     * @param {?} value
     * @return {?}
     */
    removeArray(key, value) {
        if (typeof this.table['a_' + key] !== 'undefined') {
            this.table['a_' + key].splice(this.table['a_' + key].indexOf(value), 1);
        }
        return this;
    }
    /**
     * @param {?} key
     * @return {?}
     */
    hasArray(key) {
        if (typeof this.table['a_' + key] !== 'undefined') {
            return true;
        }
        else {
            return false;
        }
    }
    /**
     * @param {?} key
     * @param {?} value
     * @return {?}
     */
    hasinArray(key, value) {
        if (typeof this.table['a_' + key] !== 'undefined') {
            if (this.table['a_' + key].indexOf(value) !== -1) {
                return true;
            }
            return false;
        }
        else {
            return false;
        }
    }
    /**
     * @return {?}
     */
    size() {
        return Object.keys(this.table).length;
    }
    /**
     * ***********************************************
     * @param {?} callback
     * @return {?}
     */
    forEach(callback) {
        for (let /** @type {?} */ key in this.table) {
            callback(key.substring(2), this.table[key]);
        }
    }
}

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */

export { HashTable };
