// angular-hashtable/bundles/angular-hashtable.umd.js
(function (global, factory) {
    typeof exports === 'object' && typeof module !== 'undefined' ? factory(exports) :
    typeof define === 'function' && define.amd ? define('angular-hashtable', ['exports'], factory) :
    (factory((global['angular-hashtable'] = {})));
}(this, (function (exports) { 'use strict';

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes} checked by tsc
     */
    /**
     * @template T, L
     */
    var /**
     * @template T, L
     */ HashTable = (function () {
        function HashTable() {
            this.table = {};
        }
        /**
         * @param {?} key
         * @param {?} value
         * @return {?}
         */
        HashTable.prototype.put = /**
         * @param {?} key
         * @param {?} value
         * @return {?}
         */
            function (key, value) {
                this.table['v_' + key] = value;
                return this;
            };
        /**
         * @param {?} key
         * @return {?}
         */
        HashTable.prototype.get = /**
         * @param {?} key
         * @return {?}
         */
            function (key) {
                return this.table['v_' + key];
            };
        /**
         * @return {?}
         */
        HashTable.prototype.getAll = /**
         * @return {?}
         */
            function () {
                var _this = this;
                var /** @type {?} */ vector = Array.from(Object.keys(this.table), function (k) { return _this.table[k]; });
                return vector;
            };
        /**
         * @return {?}
         */
        HashTable.prototype.getKeys = /**
         * @return {?}
         */
            function () {
                var /** @type {?} */ keys = Array.from(Object.keys(this.table), function (k) { return k.substring(2); });
                return keys;
            };
        /**
         * @param {?} key
         * @return {?}
         */
        HashTable.prototype.has = /**
         * @param {?} key
         * @return {?}
         */
            function (key) {
                if (typeof this.table['v_' + key] !== 'undefined') {
                    return true;
                }
                return false;
            };
        /**
         * @param {?} key
         * @return {?}
         */
        HashTable.prototype.remove = /**
         * @param {?} key
         * @return {?}
         */
            function (key) {
                delete this.table['v_' + key];
                return this;
            };
        /**
         * @param {?} key
         * @param {?} value
         * @return {?}
         */
        HashTable.prototype.putArray = /**
         * @param {?} key
         * @param {?} value
         * @return {?}
         */
            function (key, value) {
                if (typeof this.table['a_' + key] === 'undefined') {
                    this.table['a_' + key] = [];
                }
                this.table['a_' + key].push(value);
                return this;
            };
        /**
         * @param {?} key
         * @return {?}
         */
        HashTable.prototype.getArray = /**
         * @param {?} key
         * @return {?}
         */
            function (key) {
                if (typeof this.table['a_' + key] === 'undefined') {
                    this.table['a_' + key] = [];
                }
                return this.table['a_' + key];
            };
        /**
         * @param {?} key
         * @param {?} value
         * @return {?}
         */
        HashTable.prototype.removeArray = /**
         * @param {?} key
         * @param {?} value
         * @return {?}
         */
            function (key, value) {
                if (typeof this.table['a_' + key] !== 'undefined') {
                    this.table['a_' + key].splice(this.table['a_' + key].indexOf(value), 1);
                }
                return this;
            };
        /**
         * @param {?} key
         * @return {?}
         */
        HashTable.prototype.hasArray = /**
         * @param {?} key
         * @return {?}
         */
            function (key) {
                if (typeof this.table['a_' + key] !== 'undefined') {
                    return true;
                }
                else {
                    return false;
                }
            };
        /**
         * @param {?} key
         * @param {?} value
         * @return {?}
         */
        HashTable.prototype.hasinArray = /**
         * @param {?} key
         * @param {?} value
         * @return {?}
         */
            function (key, value) {
                if (typeof this.table['a_' + key] !== 'undefined') {
                    if (this.table['a_' + key].indexOf(value) !== -1) {
                        return true;
                    }
                    return false;
                }
                else {
                    return false;
                }
            };
        /**
         * @return {?}
         */
        HashTable.prototype.size = /**
         * @return {?}
         */
            function () {
                return Object.keys(this.table).length;
            };
        /**
         * ***********************************************
         * @param {?} callback
         * @return {?}
         */
        HashTable.prototype.forEach = /**
         * ***********************************************
         * @param {?} callback
         * @return {?}
         */
            function (callback) {
                for (var /** @type {?} */ key in this.table) {
                    callback(key.substring(2), this.table[key]);
                }
            };
        return HashTable;
    }());

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes} checked by tsc
     */

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes} checked by tsc
     */

    exports.HashTable = HashTable;

    Object.defineProperty(exports, '__esModule', { value: true });

})));
